### PR TITLE
Allow jails to have multiple different ZFS datasets

### PR DESCRIPTION
**Pull request by cloyce**
This PR makes changes that allow jails to be given multiple separate datasets.

My use case for this is that I want to give my jails a dataset on the fast SSD storage where they're hosted, as well as a dataset on the slower bulk storage pool.  There isn't any requirement that the different datasets be on separate pools, though.

This also extends and documents the `jail_zfs_mountpoint` property so that each dataset can be assigned its own mountpoint.  It's not necessary for each dataset to have a mountpoint, though -- `none` is still the default for all datasets that don't have one explicitly set.  I think there's probably still some work to do in this area, as it's possible for a mountpoint property set or changed from within the jail to be reset to whatever is in `jail_zfs_mountpoint` when the property is modified.

The mountpoint handling requires the changes in PR #318, as the mountpoint for a particular dataset is set by doing

```
iocage set jail_zfs_mountpoint <jail> pool/dataset=/path/in/jail
```

The code that checks to see if the iocage pool should be prepended to the dataset name has been factored out into the `__zfs_dataset_add_poolname` function.

The majority of the added complexity comes from the mountpoint handling stuff.  If we'd just say that all assigned datasets start with a mountpoint of `none` and that iocage will never touch that property again, things would be a lot simpler.

Otherwise, it's mostly just a matter of changing places that dealt with `${jail_zfs_dataset}` into a loop that operates over all the values in that property.

No existing users will be affected by these changes, as whitespace is not valid in ZFS dataset names.

### CODE REVIEW

**Comment by skarekrow** [comment]
This will take longer for me to go through, as I'm hesitant to make a large change like this.

Setting multiple mountpoints on the `jail_zfs_mountpoint` property was not intended as that was meant for one parent dataset. Once the jail is running, you can create many datasets and set those to whatever mountpoints you want. 

There are a lot of formatting that needs to be done on this PR.

Also, you *can* create datasets with whitespace in them :)

```
bschneider@iocage-dev:~ % sudo zfs create "zroot/Whitespace works"
bschneider@iocage-dev:~ % zfs list
NAME                                                           USED  AVAIL  REFER  MOUNTPOINT
zroot                                                         9.13G  8.18G   192K  /zroot
zroot/Whitespace works                                         192K  8.18G   192K  /zroot/Whitespace works
```

**Comment by cloyce** [comment]
There are several issues here; I'll try to keep them separate.

### ```jail_zfs_mountpoint```

The whole point of this PR is to allow a jail to have multiple separate parent datasets.  In my case, I have jpool (2xSSD) where I want to have a parent dataset, and tank (2x2HDD) where I want to have another.  So the ```jail_zfs_mountpoint``` stuff isn't about setting mountpoints for stuff under ```jpool/zoro``` or ```tank/zoro```; it's about setting mountpoints just for those.

I think it'd be really fine if we explicitly said that the jails manage ALL the mountpoints for their delegated parent datasets.  So the only thing that iocage would do is set it to ```none``` at creation time and then never touch it again.  If we did that, we could get rid of ```jail_zfs_mountpoint``` _entirely_.

The main problem I see with getting rid of ```jail_zfs_mountpoint``` entirely is that while you can set the mountpoint from within the jail, you have to stop and start the jail to get the dataset mounted if you set the mountpoint to something directly under ```/```.  That's a goink that would be acceptable as long as it's documented.  There are of course other use cases that I'm sure require this functionality; I just can't think of them right now.

Another alternative to making ```jail_zfs_mountpoint``` per-parent-dataset would be to document a restriction that says that if you have multiple datasets that you may not use ```jail_zfs_mountpoint``` to set a mountpoint for them.  That'd ensure that behavior doesn't change for folks who only want one (and may need to set it outside the jail), and still allow me to get the capabilities I'm after.

### Formatting

I spent a lot of time on this going through the changes and making sure that they adhered to the contribution guidelines, so I'd like very much to know what lot of code needs re-formatting, and how.

There _were_ a few things that the contribution guide didn't cover.  For those I looked to the rest of the code.  The most common one for this PR was how to break a variable assignment line so that it didn't exceed the 80-character limit.  For example,
```
_local_variable=$(__some_function "${_very_long_argument1}" "${_very_long_argument2}")
```
It wasn't clear to me whether the continuations should be indented 4 spaces:
```
_local_variable=$(__some_function \
    "${_very_long_argument1}" \
    "${_very_long_argument2}")
```
 or lined up with the start of the ```$(...)``` block:
```
_local_variable=$(__some_function \
                  "${_very_long_argument1}" \
                  "${_very_long_argument2}")
```

I saw both in the code, but 4 space indentation seemed more common, so I went with that.  It also didn't hurt that there were places where lining up with the start of the function call would've made _that_ line go over 80 columns.

For that matter, I wasn't sure how hard the 80-column limit was.  There are lots of lines in the existing code that were longer, but none really egregiously so, so I couldn't tell if those were purposeful or just an oversight.  If it's okay in some circumstances to go over 80 columns sometimes, there are a few lines in the PR that would be more readable IMO.

Of course, most of this goes away if the ```jail_zfs_mountpoint``` changes also go away.

### dataset names

Mea culpa.  I thought I'd tested spaces, but I obviously only tested '!' and '|'.  (Those definitely are illegal. :)

With that said, I don't think what the delimiter is is really important.  I'd be happy to change it to something other than whitespace.  Whitespace just seems like the most natural and easy to read, and not being able to assign a dataset with whitespace in the name seems like a reasonable restriction.

I'll be happy to make any revisions necessary to get multiple parent datasets per jail.  I'm not wedded to the mountpoint stuff; I'm not going to use it.  I only implemented it because it seemed necessary for completeness of the feature.

**Comment by skarekrow** [comment]
I think if the user makes a dataset with a space in it's name, they ought to be able to figure out how to fix that mistake ;)

As you've noted, 4 space indents is the ideal method. I've been trying to change things to that when I remember to while around the code. As for what needs to be formatted, you will note that you are missing a good number of line breaks between if statements and for loops.

We also typically do this for local variables:
```
local _var1 _var2 _var3

_var1="value"
etc, etc
```
 Note the line break between the declaration and the actual assignment. You have multiple local statements is the issue, anything else that may exist in the script that does it that way is simply legacy code.

You will also note that a few of your `if` and `elif` statements lack brackets. So to cover:
```
if [ stuff ] ; then
```

Also note that you are checking output of some commands, in that instance we use `"$()"`:
```
if [ "$(echo something | sed 's/stuff/newstuff/')" ] ; then
```

On to your changes, I'm really not attached to `jail_zfs_mountpoint` either. If you wish to experiment and remove it, that's fine by me. I just want to make sure that the migration routine is also updated so any users that are on the stable releases don't have a nasty surprise when they upgrade with any jailed datasets they may be using. That should be the only reason iocage is touching the jailed dataset's mountpoint. If you find that is different, feel free to remove it. I may be forgetting something, but I feel that the compromise of starting/stopping the jail once to get this correct is a minor annoyance that if documented like you said is fine.

Appreciate all the work you're doing on this PR!

**Comment by cloyce** [comment]
Oh!  I missed cleaning up the ```local```s in the functions I added.  I see that now and will fix it.

Regarding conditionals -- I didn't forget the braces, as those are tests that don't need test(1) (which is what ```[``` is syntactic sugar for).  But if it's required, I can rewrite things like
```
if echo foobar | grep -q bar ; then
  echo found bar
fi
```
as
```
echo foobar | grep -q bar
if [ $? -eq 0 ] ; then
  echo found bar
fi
```
Is that acceptable?  I generally don't invoke test when I don't have to as it used to be an extra program invocation, but I suppose nowadays it's a built-in and doesn't cost much extra.
I think the only place where I'm doing stuff like that is for checking if a variable's value contains a string.  Is checking the exit code of grep(1) okay (as above), or do you prefer that I do something like
```
if [ "$(echo foobar | grep bar)" != "" ] ; then
```
I think that's far less clear in intent but I'm not going to argue either way.

Regarding line breaks between ```if``` and ```for``` statements, I will fix those.  I also recommend adding that to the contribution guidelines.

I'll rip out the ```jail_zfs_mountpoint``` stuff I did, but I probably won't be able to get to it for a few days.  Just want to let you know that this is important to me and I'm not ignoring it. :) Thanks for the note about the migration routine; I'll take a look at master and figure out if there's a possibility that the new changes will affect users on the stable release.  I _think_ that ```jail_zfs_mountpoint``` was introduced after the last stable release, but I'll double-check.

Thanks for answering my questions.

**Comment by skarekrow** [comment]
Sounds good!

As for the formatting:

You will find that 
```
echo foobar | grep -q bar

if [ $? -eq 0 ] ; then
  echo found bar
fi
```
 is used when we need to run an external program and then check it's exit code. When we are checking for the existence of a string, we prefer to use 
```
if [ "$(echo foobar | grep bar)" ] ; then
```

Since that will be true if the exit code is 0 and false if not. But if you want it to be more verbose,
 ```
if [ "$(echo foobar | grep bar)" != "0" ] ; then
``` 
is also acceptable, but not preferred. 

I'd like to avoid tests like ```if [ "$(echo foobar | grep bar)" != "" ] ; then``` since it's a little non-obvious why we are matching "".

I'm excited to see what you do with this as jailed datasets haven't gotten a tremendous amount of love, and a cool feature like this is excellent.

**Comment by cloyce** [comment]
I'm tired of trying to make rebase do what I think it does.

I'm going to fix all the things and make sure everything's working before trying to squash it all in again.

**Comment by skarekrow** [comment]
Yes git is a *bitch* to work with sometimes. No rush :)

**Comment by cloyce** [comment]
This PR is ready for review again.

### `jail_zfs_mountpoint`

I've simplified the `jail_zfs_mountpoint` handling.  The only change now is that if you assign multiple ZFS datasets to the jail, `jail_zfs_mountpoint` must be set to "none", and this is enforced.  Setting it to "none" doesn't change any existing mountpoint properties, so there should be no surprises for the user.

For example, assuming a pre-existing jail "`testjail`", doing something like

    # iocage set jail_zfs_dataset=tm/testjailds1 testjail
    # iocage set jail_zfs_mountpoint=/zfs testjail
    # iocage set jail_zfs=on testjail
    # iocage start testjail
    * Starting 51d99ca7-f137-11e5-ab92-d05099264f2e (testjail)
      + Started (shared IP mode) OK
      + Starting services        OK
    # iocage exec testjail zfs list
    NAME             USED  AVAIL  REFER  MOUNTPOINT
    tm              1.22T   552G   144K  none
    tm/testjailds1    96K   552G    96K  /zfs

will result in the `tm/testjailds1` dataset being created and mounted on `/zfs` inside the jail.  So far so good.  If the user then adds another dataset with:

    # iocage set jail_zfs_dataset="tm/testjailds1 tm/testjailds2" testjail
    NOTICE: The setting for jail_zfs_mountpoint must be 'none' when multiple
            datasets are assigned to the jail.

iocage does add the new dataset and sets `jail_zfs_mountpoint` to "none":

    # iocage get all testjail | grep zfs
    allow_mount_zfs:1
    jail_zfs:on
    jail_zfs_dataset:tm/testjailds1 tm/testjailds2
    jail_zfs_mountpoint:none

However, the mountpoint that was set for `tm/testjailds1` is not changed:

    # iocage exec testjail zfs list
    NAME             USED  AVAIL  REFER  MOUNTPOINT
    tm              1.22T   552G   144K  none
    tm/testjailds1    96K   552G    96K  /zfs
    tm/testjailds2    96K   552G    96K  none

Even in the single dataset case, if the user sets a mountpoint directly via `zfs set` that's different from the setting in `jail_zfs_mountpoint`, iocage will no longer touch it.

There's no special handling for this required for migration because `jail_zfs_mountpoint` doesn't exist in master.

### Dataset creation refactored

There were a couple of places in the code where assigned datasets could be created.  They were almost identical (`zfs create` called from `__create_jail` got a `-u`; otherwise it didn't), so I factored those out into a separate function in `lib/ioc-zfs`.  This allows more consistent behavior -- datasets get created when you turn `jail_zfs` on (either at jail creation time, or by setting the property later) and now also when you set the `jail_zfs_dataset` property.

The code that figures out if `${pool}/iocage/jails` needs to be prepended to a dataset name has also been factored out into a separate function; that code was duplicated in several places in `ioc-rc` as well as `ioc-common` and `ioc-zfs`.

### Otherwise...

Otherwise, the changes are mainly about turning places that assumed that `jail_zfs_dataset` was the name of one and only one dataset into loops that operate on all the datasets in the list.

File: lib/ioc-common [hunk 1]
```diff
@@ -94,30 +107,67 @@ __set_jail_prop () {
...
+
+            # Enforce the rule that jail_zfs_mountpoint must be "none" when
+            # the jail has multiple datasets assigned.
+            if [ "$(echo "${_jail_zfs_datasets}" | grep '[ 	]')" ] ; then
+                if [ "${jail_zfs_mountpoint}" != "none" ] ; then
+                    # Do not __die here, as this may be called from
+                    # __create_jail, and that could leave the jail incompletely
+                    # configured.  Instead, just force the value to "none" and
+                    # carry on.
+                    echo "NOTICE: With multiple ZFS datasets per jail," \
```
Comment (skarekrow) [inline lib/ioc-common]: Use `__info` instead of `NOTICE:`.

File: lib/ioc-zfs [hunk 2]
```diff
@@ -1328,3 +1330,74 @@ __send () {
...
+    if [ -z "${_jail_zfs_mountpoint}" ] ; then
+        _jail_zfs_mountpoint="$(__get_jail_prop jail_zfs_mountpoint \
+            "${_uuid}" "${_dataset}")"
+    fi
+
+    # Check to make sure that jail_zfs_mountpoint is "none" if there are
+    # multiple datasets assigned to the jail
+    if [ "${_jail_zfs_mountpoint}" != "none" ] ; then
+        if [ "$(echo "${_jail_zfs_datasets}" | grep "[ 	]")" ] ; then
+            echo "NOTICE: The setting for jail_zfs_mountpoint must be 'none'" \
```
Comment (skarekrow) [inline lib/ioc-zfs]: Use `__info` instead of `NOTICE:`.

File: lib/ioc-zfs [hunk 3]
```diff
@@ -1328,3 +1330,74 @@ __send () {
...
+                 "when multiple"
+            echo "        datasets are assigned to the jail."
```
Comment (skarekrow) [inline lib/ioc-zfs]: Unnecessary echo perhaps?

**Comment by skarekrow** [comment]
Looks great, thanks! Made some comments, more styling then anything.

**Comment by cloyce** [comment]
Changed 'echo "NOTICE:...' to `__info`.

The `echo` wasn't extra; I was wrapping the message at 80 characters.  I shortened the message (made it match the other place where it can be emitted), so that shouldn't be an issue now.

And... oh man.  "A little love" was quite the understatement.  I'll try to have conflicts resolved by Friday.

**Comment by skarekrow** [comment]
Sounds good! And hmm I didn't recall touching much of the `jail_zfs` stuff. Sorry man! Heh, but it's for the greater good. The tool is much faster now.

**Comment by cloyce** [comment]
Well, I think I'm done with this now.  I hope someone finds this useful.

**Comment by skarekrow** [comment]
:+1:

**Pull request merged**